image/manifest.go

```diff
@@ -26,28 +26,269 @@ import (
+		// Found a rule for this key
+		if err := allowed.Check(used); err != nil {
+			return err
+		}
+	}
+	if _, ok := sm.used[key]; !ok {
+		sm.used[key] = used
+	}
+	return nil
+}
+
+// SetAllowedSnapRevision adds a revision rule for the given snap name, meaning
+// that any snap marked used through MarkSnapRevisionUsed will be validated against
+// this rule. The manifest will only allow one revision per snap, meaning that any
+// subsequent calls to this will be ignored.
+func (sm *SeedManifest) SetAllowedSnapRevision(snapName string, revision int) error {
+	if revision == 0 {
+		return fmt.Errorf("cannot add a rule for a zero-value revision")
+	}
+	sm.addAllowedOnce(&SeedManifestSnapRevision{
+		SnapName: snapName,
+		Revision: snap.R(revision),
+	})
+	return nil
+}
+
+// SetAllowedValidationSet adds a sequence rule for the given validation set, meaning
+// that any validation set marked for use through MarkValidationSetUsed must match the
+// given parameters. The manifest will only allow one sequence per validation set,
+// meaning that any subsequent calls to this will be ignored.
+func (sm *SeedManifest) SetAllowedValidationSet(accountID, name string, sequence int, pinned bool) error {
+	if sequence <= 0 {
+		return fmt.Errorf("cannot add allowed validation set for a unknown sequence")
+	}
+	sm.addAllowedOnce(&SeedManifestValidationSet{
+		AccountID: accountID,
+		Name:      name,
+		Sequence:  sequence,
+		Pinned:    pinned,
+	})
+	return nil
+}
+
+// MarkSnapRevisionUsed validates the revision of the given snap name against any
+// previously setup revision rules by SetAllowedSnapRevision. Attempting to mark
+// the same snap of multiple revisions used will be ignored..
+func (sm *SeedManifest) MarkSnapRevisionUsed(snapName string, revision int) error {
+	return sm.addUsedOnce(&SeedManifestSnapRevision{
+		SnapName: snapName,
+		Revision: snap.R(revision),
+	})
+}
+
+// MarkValidationSetUsed tracks the used validation set in the manifest.T he manifest
+// will only allow one revision per snap, meaning that any subsequent calls to this will
+// be ignored.
+func (sm *SeedManifest) MarkValidationSetUsed(accountID, name string, sequence int, pinned bool) error {
+	if sequence <= 0 {
+		return fmt.Errorf("cannot mark validation-set \"%s/%s\" used, sequence must be set", accountID, name)
+	}
+	return sm.addUsedOnce(&SeedManifestValidationSet{
+		AccountID: accountID,
+		Name:      name,
+		Sequence:  sequence,
+		Pinned:    pinned,
+	})
+}
+
+// AllowedRevision retrieves any specified revision rule for the snap
+// name.
+func (sm *SeedManifest) AllowedSnapRevision(snapName string) snap.Revision {
+	if allowed, ok := sm.allowed[snapName]; ok {
+		if sr, ok := allowed.(*SeedManifestSnapRevision); ok {
+			return sr.Revision
+		}
+	}
+	return snap.Revision{}
+}
+
+// ValidationSetsAllowed returns the validation sets specified as allowed.
+func (sm *SeedManifest) ValidationSetsAllowed() []*SeedManifestValidationSet {
```

Comment:
for symmetry this should be called AllowedValidationSets, it's not used yet? but we will need to add a way to feed this to the Writer to use the sequences when specified and the model itself doesn't pin

Reply:
Fixed, good point on being consistent in the naming! It's not used yet but will be used by the seedwriter as you mentioned.